dsp/dsp.py

```diff
@@ -0,0 +1,254 @@
+    assert(int(padding_factor) == padding_factor)
+    
+    zero_phase = config['zero_phase']
+    sqrt_window = config['sqrt_window']
+
+    N = len(signal)
+    H = hop_size
+    num_hops = (N - window_size) // H + 1
+    M = num_hops * H + window_size
+    if M > N:
+        signal = np.append(signal, np.zeros(M - N))
+    assert(len(signal) == M)
+    
+    num_hops = (M - window_size) // H + 1
+    
+    if sqrt_window and np.all(window > 0):
+        window = np.sqrt(window)
+
+    frame_indexer = np.arange(window_size).reshape(1, -1) + hop_size * np.arange(num_hops).reshape(-1, 1)
+    
+    windowed_frames = np.copy(signal[frame_indexer]) * window
+
+    frame_size = get_windowed_frame_size(config)
+    padding_size = frame_size - window_size
+    if padding_size > 0:
+        windowed_frames = np.append(windowed_frames, np.zeros((num_hops, padding_size)), axis=1)
+
+    if zero_phase:
+        zero_phase_shift = 1 - window_size // 2
+        windowed_frames = np.roll(windowed_frames, zero_phase_shift, axis=1)
+    
+    windowed_frame_data = {
+        'padded_duration_samp': M,
+        'frame_size': frame_size,
+        'frames': windowed_frames
+    }
+        
+    if config['return_padded_signal']:
+        windowed_frame_data['padded_signal'] = signal
+    return windowed_frame_data
+
+def windowed_frame_synthesis(windowed_frame_data, config):
+    windowed_frames = windowed_frame_data['frames']
+    M = windowed_frame_data['padded_duration_samp']
+    
+    window = config['window']
+    window_size = config['window_size']
+    assert(len(window) == window_size)
+    hop_size = config['hop_size']
+    assert(hop_size <= window_size)
+    padding_factor = config['padding_factor']
+    assert(int(padding_factor) == padding_factor)
+    
+    zero_phase = config['zero_phase']
+    sqrt_window = config['sqrt_window']
+    
+    signal = np.zeros(M)
+    
+    if zero_phase:
+        zero_phase_shift = 1 - window_size // 2
+        windowed_frames = np.roll(windowed_frames, -zero_phase_shift, axis=1)
+    
+    frame_size = get_windowed_frame_size(config)
+    assert(frame_size == windowed_frame_data['frame_size'])
+    padding_size = frame_size - window_size
+    if padding_size > 0:
+        windowed_frames = windowed_frames[:, :-padding_size]
+        
+    if sqrt_window and np.all(window > 0):
+        window = np.sqrt(window)
+        windowed_frames[:, : window_size] *= window
+
+    num_hops = windowed_frames.shape[0]    
+    for i in range(num_hops):
+        n_start = i * hop_size
+        n_end = n_start + window_size
+        signal[n_start : n_end] += windowed_frames[i]
+    return signal
+
+def get_fft_bin_phase_frequencies(num_bins, hop_size, sample_rate):
+    bin_frequencies = np.linspace(0, sample_rate // 2, num=num_bins, endpoint=True)
+    bin_phases = np.linspace(0, np.pi * hop_size, num=num_bins, endpoint=True)
+    return bin_frequencies, bin_phases
+
+def get_feature_diffs(feature_frames):
```

Comment:
An implicit assumption in this and other functions here is that `time` corresponds to `axis=0`. Maybe document this.